nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -127,28 +127,47 @@ BtNavigator::on_configure(const rclcpp_lifecycle::State & /*state*/)
   blackboard_->set<int>("number_recoveries", 0);  // NOLINT
 
   // Get the BT filename to use from the node parameter
-  std::string bt_xml_filename;
-  get_parameter("bt_xml_filename", bt_xml_filename);
+  get_parameter("default_bt_xml_filename", default_bt_xml_filename_);
+
+  if (!loadBehaviorTree(default_bt_xml_filename_)) {
+    RCLCPP_ERROR(get_logger(), "Error loading XML file: %s", default_bt_xml_filename_.c_str());
+    return nav2_util::CallbackReturn::FAILURE;
+  }
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool
+BtNavigator::loadBehaviorTree(const std::string & bt_xml_filename)
+{
+  // Use previous BT if it is the existing one
+  if (bt_xml_filename != "" && current_bt_xml_filename_ == bt_xml_filename) {
+    return true;
+  }
 
   // Read the input BT XML from the specified file into a string
   std::ifstream xml_file(bt_xml_filename);
 
   if (!xml_file.good()) {
     RCLCPP_ERROR(get_logger(), "Couldn't open input XML file: %s", bt_xml_filename.c_str());
-    return nav2_util::CallbackReturn::FAILURE;
+    return false;
   }
 
-  xml_string_ = std::string(
+  auto xml_string = std::string(
     std::istreambuf_iterator<char>(xml_file),
     std::istreambuf_iterator<char>());
 
   RCLCPP_DEBUG(get_logger(), "Behavior Tree file: '%s'", bt_xml_filename.c_str());
-  RCLCPP_DEBUG(get_logger(), "Behavior Tree XML: %s", xml_string_.c_str());
+  RCLCPP_DEBUG(get_logger(), "Behavior Tree XML: %s", xml_string.c_str());
+
+  // Create new BT to ensure it is a clean BT
+  tree_ = std::make_shared<BT::Tree>();
 
   // Create the Behavior Tree from the XML input
-  tree_ = bt_->buildTreeFromText(xml_string_, blackboard_);
+  *tree_ = bt_->buildTreeFromText(xml_string, blackboard_);
```

Comment:
I don't see why this requires a pointer since you're casting it into the pointer memory. What specifically went wrong when this was a stack variable?

Reply:
I have returned tree_ to the stack. Now I don't see the problem.